### PR TITLE
Add CI check to test macOS experimental setup

### PR DESCRIPTION
**Pull request by lbianchi-lbl**
## Summary/Motivation:
- Test viability of known workarounds for some degree of partial and/or experimental macOS support for WaterTAP

## Changes proposed in this PR:
- Add CI job running the WaterTAP test suite in macOS

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by bknueven** [comment]
If we merge #283 I only see one failure on my mac locally:
```
=================================================================== short test summary info ===================================================================
FAILED unit_models/tests/test_reverse_osmosis_1D.py::TestReverseOsmosis::test_var_scaling - AssertionError: assert [(<pyomo.core...36105133e-07)] == []
==================================================== 1 failed, 547 passed, 8 skipped in 292.93s (0:04:52) =====================================================
```
Since all the other 1DRO tests pass, and this is just checking the variable scaling, maybe it should be disabled.

**Comment by lbianchi-lbl** [comment]
@bknueven That's good to know, thanks. If the (known) failures are limited to that single test, we could also add a `pytest.mark.skipif()` or `pytest.mark.xfail()` to it, either using e.g. `"darwin" in sys.platform` or (possibly, technically more correct?) checking if the Ipopt executable is the one from `idaes get-extension` or the one from `conda-forge`.

File: .github/workflows/checks.yml [hunk 1]
```diff
@@ -187,3 +191,44 @@ jobs:
...
+        pip show pyomo idaes-pse
+        echo '::endgroup::'
+    - name: Install Ipopt from conda-forge
+      run:
+        conda install --quiet --yes -c conda-forge ipopt
+    - name: Build Pyomo extensions
+      run: |
+        conda install --quiet --yes cmake
+        # some failures are expected, but this should succeed as long as pynumero is built correctly
+        pyomo build-extensions || grep '.*OK.*pynumero'
```
Comment (bknueven) [inline .github/workflows/checks.yml]:
```suggestion
        pyomo build-extensions | grep '.*OK.*pynumero'
```

File: .github/workflows/checks.yml [hunk 2]
```diff
@@ -187,3 +191,44 @@ jobs:
...
+        pip show pyomo idaes-pse
+        echo '::endgroup::'
+    - name: Install Ipopt from conda-forge
+      run:
+        conda install --quiet --yes -c conda-forge ipopt
+    - name: Build Pyomo extensions
+      run: |
+        conda install --quiet --yes cmake
+        # some failures are expected, but this should succeed as long as pynumero is built correctly
+        pyomo build-extensions | grep '.*OK.*pynumero'
```
Comment (bknueven) [inline .github/workflows/checks.yml]:
```suggestion
        pyomo build-extensions
        python -c "from pyomo.contrib.pynumero.asl import AmplInterface; exit(0) if AmplInterface.available() else exit(1)"
```

File: .github/workflows/checks.yml (hunk 1 above)
Comment (lbianchi-lbl) [reply .github/workflows/checks.yml]: @bknueven thanks for catching that!

File: .github/workflows/checks.yml [hunk 3]
```diff
@@ -187,3 +191,45 @@ jobs:
...
+        echo '::endgroup::'
+    - name: Install Ipopt from conda-forge
+      run:
+        conda install --quiet --yes -c conda-forge ipopt
+    - name: Build Pyomo extensions
+      run: |
+        conda install --quiet --yes cmake
+        # some failures are expected, but this should succeed as long as pynumero is built correctly
+        pyomo build-extensions
+        python -c "from pyomo.contrib.pynumero.asl import AmplInterface; exit(0) if AmplInterface.available() else exit(1)"
```
Comment (bknueven) [inline .github/workflows/checks.yml]:
```suggestion
        pyomo build-extensions || python -c "from pyomo.contrib.pynumero.asl import AmplInterface; exit(0) if AmplInterface.available() else exit(1)"
```

**Review by bknueven — DISMISSED** [review]
Somebody else should review the macOS installation instructions.

File: docs/getting_started/install.rst [hunk 4]
```diff
@@ -62,12 +62,13 @@ If you are going to use WaterTAP's functionality, but *do not* plan to contribut
...
-Installing solvers distributed through IDAES Extensions
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Installing solvers
+^^^^^^^^^^^^^^^^^^
 
-After installing WaterTAP, the ``idaes get-extensions`` command can be used to automatically install the solvers distributed as part of the IDAES Extensions.
+Windows and Linux Users: solvers distributed through IDAES Extensions
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 
-.. important:: Depending on your operating system, additional steps might be needed. For more information, refer to the `IDAES installation guide <https://idaes-pse.readthedocs.io/en/stable/getting_started/index.html>`_.
```
Comment (lbianchi-lbl) [inline docs/getting_started/install.rst]: Should we consider keeping this line? AFAIK the need for additional steps before `idaes get-extensions` (e.g. on Linux, installing some system packages) still applies.

File: .github/workflows/checks.yml [hunk 5]
```diff
@@ -187,3 +187,43 @@ jobs:
...
+        pip show pyomo idaes-pse
+        echo '::endgroup::'
+    - name: Install Ipopt from conda-forge
+      run:
+        conda install --quiet --yes -c conda-forge ipopt
+    - name: Build Pyomo extensions
+      run: |
+        conda install --quiet --yes cmake
+        # some failures are expected, but this should succeed as long as pynumero is built correctly
+        pyomo build-extensions || python -c "from pyomo.contrib.pynumero.asl import AmplInterface; exit(0) if AmplInterface.available() else exit(1)"
```
Comment (lbianchi-lbl) [inline .github/workflows/checks.yml:225]: Nice - that's a much better way to check if `pyomo build-extensions` is successful than trying to parse its output.

**Review by lbianchi-lbl — COMMENTED** [review]
I'm unable to approve (probably because I created the PR) but I like @bknueven's changes to the GHA workflow and RTD instructions.

File: docs/getting_started/install.rst [hunk 6]
```diff
@@ -62,12 +62,13 @@ If you are going to use WaterTAP's functionality, but *do not* plan to contribut
...
 .. _install-idaes-ext:
 
-Installing solvers distributed through IDAES Extensions
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Installing solvers
+^^^^^^^^^^^^^^^^^^
 
-After installing WaterTAP, the ``idaes get-extensions`` command can be used to automatically install the solvers distributed as part of the IDAES Extensions.
+Windows and Linux Users: solvers distributed through IDAES Extensions
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```
Comment (bknueven) [inline docs/getting_started/install.rst:70]:
```suggestion
.. important:: Depending on your operating system, additional steps might be needed. For more information, refer to the `IDAES installation guide <https://idaes-pse.readthedocs.io/en/stable/getting_started/index.html>`_.
```

File: docs/getting_started/install.rst (hunk 4 above)
Comment (bknueven) [reply docs/getting_started/install.rst]: I agree -- I put it back in.

File: .github/workflows/checks.yml (hunk 5 above)
Comment (bknueven) [reply .github/workflows/checks.yml:225]: For some reason pipe was still failing, even though it worked on my system. GHA might be checking the output of each command when pipe is used.

At any rate, I do like the direct import checking better as well.

**Review by bknueven — DISMISSED** [review]

**Comment by ksbeattie** [comment]
I just ran through the new steps in the install section and it all seemed to work.  Two ideas:
- After running `pyomo build-extensions` the user doesn't need `cmake` anymore, right?  Perhaps an optional `conda uninstall cmake` could be suggested?
- It wasn't clear to me, just following the steps, if things were any better than they were before running `pyomo build-extensions`, i.e. `python -c "from pyomo.contrib.pynumero.asl import AmplInterface; print(AmplInterface.available())"` prints `True` even in a conda env where I **haven't** run `pyomo build-extensions`

**Comment by bknueven** [comment]
> - After running `pyomo build-extensions` the user doesn't need `cmake` anymore, right?  Perhaps an optional `conda uninstall cmake` could be suggested?

This is correct -- I can add that suggestion.


> - It wasn't clear to me, just following the steps, if things were any better than they were before running `pyomo build-extensions`, i.e. `python -c "from pyomo.contrib.pynumero.asl import AmplInterface; print(AmplInterface.available())"` prints `True` even in a conda env where I **haven't** run `pyomo build-extensions`

Ah, good observation. Pyomo places the built extensions in `~/.pyomo`, so they can be picked up by other environments. If you remove this directory you'll see this print `False`.

I suppose this could be an issue if a user is using pynumero and running different versions of Pyomo in different environments, but we probably don't need to worry about such a user.

**Review by bknueven — APPROVED** [review]

**Comment by bknueven** [comment]
@ksbeattie I think this is ready-to-go now. I added notes in the documentation to address both your concerns above.

**Review by ksbeattie — APPROVED** [review]
LGTM

**Pull request merged**